### DISCUSSION TITLE
How to read older versions of the docs?

### DISCUSSION THREAD

**Discussion by amudh** [python-pillow/Pillow#6894] [Q&A] [answered]
For a project, I need a specific version of the documentation. E.g., 8.1.0. Only the latest and stable versions are available, and other versions are disabled/removed. I went to this link directly: https://readthedocs.org/projects/pillow/versions/ and could only see those versions. Alternatively, I went to the `docs` folder on the repo but could only see the source of the documentation in `.rst` file formats.

How to get older versions? Meanwhile, I can refer to the latest version to get started. I would appreciate any help.

**Comment by Yay295** [comment]
It looks like other versions need to be manually activated. https://docs.readthedocs.io/en/stable/tutorial/#modifying-versions

**Reply by nulano** [reply]
IIRC it was a deliberate choice to deactivate old versions to improve SEO: #4549

**Reply by aclark4life** [reply]
Copy that, disabled 8.1.0 again.

**Reply by amudh** [reply]
I am not clear on the exact problems from issue link #4549. But, getting access to an older version of the documentation is a fairly common requirement. For example, Python documentation supports multiple versions. Is this a limitation of `readthedocs.io`?

**Reply by aclark4life** [reply]
> getting access to an older version of the documentation is a fairly common requirement.

I don't know about requirement, but yes it's nice to have old versions available.

> Is this a limitation of `readthedocs.io`?

It's up to us to decide what we want to do and RTD supports developers' efforts as best they can.

**Reply by adelplanque** [reply]
It would be more respectful to users to keep older versions of the documentation. I'm currently migrating code from version 6.2 to the latest version. Removing access to older versions makes it harder.

**Reply by aclark4life** [reply]
@adelplanque What do you mean by "removing access"? [6.2 is right here, enjoy](https://files.pythonhosted.org/packages/87/dc/7597336c48796d4a836007460148b7baf7f278ad42b73d49047eb0e8194c/Pillow-6.2.0.tar.gz)!

**Reply by adelplanque** [reply]
This isn't about the package, but about the documentation. I had to rebuild the 6.2 documentation, which was complex since many dependencies had changed.
My code to migrate included a deleted function, and the work is still easier when you have the old documentation.
It would be nice, out of respect for users, to put the old documentation back online, as is often the case.

**Reply by aclark4life** [reply]
I don't disagree, but it looks like RTD is having trouble with it too.

<img width="1205" alt="Screenshot 2025-05-02 at 1 38 32 PM" src="https://github.com/user-attachments/assets/a02cbb2b-86dc-4776-ad82-02d182ecffbc" />

**Reply by hugovk** [reply]
6.2 was released six years ago, I'm not surprised if they're not building due to dependencies. However, the docs are reStructuredText, so should be quite readable in raw form as well. If you have any specific questions about migrating from 6 to 11, please feel free to open a new issue or thread.

**Reply by aclark4life** [reply]
That's true, here's the [6.2 branch docs](https://github.com/python-pillow/Pillow/blob/6.2.x/docs/index.rst). Also I'd be surprised if there was anything in those docs that actually helped move from 6.2 to 11 that isn't already in the [release notes for the current docs](https://pillow.readthedocs.io/en/stable/releasenotes/index.html).

**Reply by adelplanque** [reply]
The function in question is ImageDraw.textsize. Its removal is indeed mentioned in the release notes for version 10.0.0 (there are over 30 versions between 6.2.2 and 11.2.1) and a note is dedicated to it in the depreciations page.
I'm used to migrating legacy code; when a regression test reveals an API change, the most practical way is to look for the function in the old API documentation to understand its purpose, and look for an alternative in the new documentation.
Due to my legacy code, I still have python 2.7 environments, which is what allows me to rebuild the old doc, but it's still painful.

**Comment by radarhere** [comment, marked as answer]
If you would like, you can generate those versions of the documentation on your machine.

```
git clone git@github.com:python-pillow/Pillow
cd Pillow
git checkout 9.2.0
python3 -m pip install --force-reinstall .
make doc
open docs/_build/html/index.html
```

**Reply by amudh** [reply]
This seems to have worked. Thank you! Although there were couple of issues. One related to `imageio` module being incompatible with `8.1.0` tag. Since I only want the docs at this point, I ignored that. However for the `make doc` command to succeed, I had to manually install the following
```
pip install sphinx_rtd_theme
pip install sphinx_issues
pip install sphinx_removed_in
```

I now have the docs running locally.

**Comment by hugovk** [comment]
The .rst files should be fairly readable via the tag on GitHub:

https://github.com/python-pillow/Pillow/tree/8.1.0/docs

**Comment by aclark4life** [comment]
We can also manually activate older versions if need be, e.g. https://pillow.readthedocs.io/en/8.1.0/

**Reply by amudh** [reply]
Wow, this was easy. Thanks!

**Reply by SubaruArai** [reply]
I don't think this is working anymore

**Reply by radarhere** [reply]
This specific link was later disabled. See https://github.com/python-pillow/Pillow/discussions/6894#discussioncomment-4729249
